src/Microsoft.ML.Data/DataLoadSave/DataOperationsCatalog.cs

```diff
@@ -491,50 +491,87 @@ internal static IEnumerable<TrainTestData> CrossValidationSplit(IHostEnvironment
                 {
-                    var origStratCol = samplingKeyColumn;
-                    samplingKeyColumn = data.Schema.GetTempColumnName(samplingKeyColumn);
+                    var hashInputColumnName = samplingKeyColumn;
                     // HashingEstimator currently handles all primitive types except for DateTime, DateTimeOffset and TimeSpan.
                     var itemType = type.GetItemType();
                     if (itemType is DateTimeDataViewType || itemType is DateTimeOffsetDataViewType || itemType is TimeSpanDataViewType)
-                        data = new TypeConvertingTransformer(env, origStratCol, DataKind.Int64, origStratCol).Transform(data);
+                    {
+                        data = new TypeConvertingTransformer(env, newSamplingKeyColumn, DataKind.Int64, samplingKeyColumn).Transform(data);
+                        hashInputColumnName = newSamplingKeyColumn;
+                    }
 
-                    var localSeed = seed.HasValue ? seed : ((ISeededEnvironment)env).Seed.HasValue ? ((ISeededEnvironment)env).Seed : null;
                     var columnOptions =
-                        localSeed.HasValue ?
-                        new HashingEstimator.ColumnOptions(samplingKeyColumn, origStratCol, 30, (uint)localSeed.Value, combine: true) :
-                        new HashingEstimator.ColumnOptions(samplingKeyColumn, origStratCol, 30, combine: true);
+                        seedToUse.HasValue ?
+                        new HashingEstimator.ColumnOptions(newSamplingKeyColumn, hashInputColumnName, 30, (uint)seedToUse.Value, combine: true) :
+                        new HashingEstimator.ColumnOptions(newSamplingKeyColumn, hashInputColumnName, 30, combine: true);
                     data = new HashingEstimator(env, columnOptions).Fit(data).Transform(data);
                 }
                 else
                 {
-                    if (!data.Schema[samplingKeyColumn].IsNormalized() && (type == NumberDataViewType.Single || type == NumberDataViewType.Double))
+                    if (data.Schema[samplingKeyColumn].IsNormalized() || (type != NumberDataViewType.Single && type != NumberDataViewType.Double))
+                    {
+                        data = new ColumnCopyingEstimator(env, (newSamplingKeyColumn, samplingKeyColumn)).Fit(data).Transform(data);
+                    }
+                    else
                     {
-                        var origStratCol = samplingKeyColumn;
-                        samplingKeyColumn = data.Schema.GetTempColumnName(samplingKeyColumn);
-                        data = new NormalizingEstimator(env, new NormalizingEstimator.MinMaxColumnOptions(samplingKeyColumn, origStratCol, ensureZeroUntouched: true)).Fit(data).Transform(data);
+                        data = new NormalizingEstimator(env, new NormalizingEstimator.MinMaxColumnOptions(newSamplingKeyColumn, samplingKeyColumn, ensureZeroUntouched: false)).Fit(data).Transform(data);
                     }
```

Comment:
Notice the change of `ensureZeroUntouched` from `true` to `false`.

I did this for the reasons you've mentioned me offline, @yaeldekel , about another issue that would happen if the values of the samplingKeyColumn weren't inside 0-1. In fact, one of the tests I've added on this PR won't pass without `ensureZeroUntouched = false`.

Reply:
I agree that this change is needed. In this case there is no down side to it, since we know that the column being normalized is not a vector column (otherwise it would have gone into the other branch of the "if").
I think we should also consider normalizing even if `data.Schema[samplingKeyColumn].IsNormalized()` is true, because it could happen that it will be true but the values won't be in the range 0-1 (for example, if `samplingKeyColumn` was normalized using `MinMaxNormalizer` with `ensureZeroUntouched=true`).

---
In reply to: [441994977](https://github.com/dotnet/machinelearning/pull/5227#discussion_r441994977) [](ancestors = 441994977)

Reply:
Updated so that it normalizes even if IsNormalized is true.

---
In reply to: [443435362](https://github.com/dotnet/machinelearning/pull/5227#discussion_r443435362) [](ancestors = 443435362,441994977)